### PR TITLE
Add support for webpack's CommonsChunkPlugin

### LINKED ISSUE #253: A way to stop loading the same module multiple times.
Have a look at the following two pages:

**pages/index.js**

```jsx
import React from 'react'
import Counter from '../components/Counter'
import Link from 'next/link'

console.log("Home", Counter());

export default () => (
  <div>
    <p>Home</p>
    <Link href="/about" />
  </div>
)
```

**pages/about.js**

```jsx
import React from 'react'
import Counter from '../components/Counter'

console.log("About", Counter());

export default () => (
  <div>
    <p>About</p>
  </div>
)
```

And here's the Counter component

**components/Counter.js**

```jsx
console.log('Counter');

let counter = 0;

export default () => {
  return ++counter;
}
```

Here's the situation:

* Visit "/" - Then the `Counter` module gets loads for the first time
* Then I visit "/about" by clicking the link on the home page - Then the `Counter` module loads again.

This gives us two issues:

* Loading common code more than once. 
* Our app might not work properly. (Here our counter is not gonna work properly as there are two such modules)

### Possible solution

1. Allow the developer to define a set of modules as common modules.
2. Then bundle them into a separate output and load with every request.
3. We also need to add them to externals in the webpack config.

### PR DESCRIPTION
**Pull request by arunoda**
Fixes #253
Similar to https://github.com/zeit/next.js/pull/286 (We didn't take #286 in)

Here now common modules loads once using the webpack's CommonsChunkPlugin. Still, we use Next.js module loading system and SSR features.

As a result of this, we don't need to worry about loading the next-bundle and even creating it.

### CODE REVIEW

File: examples/shared-modules/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Example app using shared modules
+
+This example features:
+
+* An app with two pages which has a common Counter component
+* That Counter component maintain the counter inside it's module.
```
Comment (rauchg) [inline examples/shared-modules/README.md]: `its`

Comment (rauchg) [inline examples/shared-modules/README.md]: Maybe reword this sentence? It's not 100% clear to me

Comment (rauchg) [inline examples/shared-modules/README.md]: Maybe say something along the lines: the common plugin maintains state that is initialized only once.

**Review by rauchg — CHANGES_REQUESTED** [review]
looking great!!!

**Comment by nkzawa** [comment]
👍

**Comment by arunoda** [comment]
Superb!

**Comment by jonaswindey** [comment]
Does this mean no configuration is required to make a component/module "common"?

**Comment by arunoda** [comment]
@jonaswindey no we don't need. 

Currently it uses defaults options for [commonChunksPlugin](https://webpack.github.io/docs/list-of-plugins.html#commonschunkplugin)

In a future(near) version of Next.js we'll offer some ways to configure options for `commonChunksPlugin` and move some specific modules into the common chunk.

**Comment by arunoda** [comment]
@jonaswindey more info about the current logic of common modules.

If a module used in all of the pages, it will move into the commons JavaScript bundle.
So, individual pages doesn't contain that modules.
If not, individual modules will contain a copy of that module.

But in anyway, a module only loads once even if pages contains multiple copies of it.

**Comment by adamrights** [comment]
One thing to consider is how aggressively assets are concatenated in http2 setups:

> Concatenation is not necessarily an anti-pattern with HTTP/2. Under HTTP/2, it’s good practice to keep individual files small and ensure that resources are only served when needed. That being said, other factors can affect the the speed tradeoffs of individual resources. For example, when Khan Academy served over 300 individual JavaScript files to HTTP/2 users, they saw a degradation in performance due to less efficient compression over multiple files, and server delays related to reading each file from disk. For more information, see Khan Academy’s article and Smashing Magazine’s HTTP/2 guide.
> 
> Generally speaking, organizing your resources into smaller, logical files rather than bundling them one large file offers the best performance over HTTP/2. The number of files that you can serve over HTTP/2 for a single URL without degrading performance depends heavily on the codebase and the server. We suggest keeping the number of files below 50 per URL, as that seems to be the point at which many servers suffer from having to read so many individual files from disk.

[source](https://designsystem.digital.gov/performance/http2/)
